src/StreamJsonRpc.Tests/DuplexPipeMarshalingTests.cs

```diff
@@ -292,17 +294,59 @@ public async Task ClientCanSendWriteOnlyStreamToServer(bool orderedArguments)
         await this.AssertStreamClosesAsync(duplexStream.Item2);
     }
 
-    [Fact(Skip = "This would require JsonRpc to hard-code understanding of formatter tricks, or more design work.")]
-    public void ServerMethodThatReturnsIDuplexPipeIsRejected()
+    [Fact]
+    public async Task ServerMethodThatReturnsDuplexPipe()
     {
-        this.serverRpc.AllowModificationWhileListening = true;
-        Assert.Throws<NotSupportedException>(() => this.serverRpc.AddLocalRpcTarget(new ServerWithIDuplexPipeReturningMethod()));
+        byte[] buffer = new byte[32];
+        Stream result = await this.clientRpc.InvokeAsync<Stream>(nameof(Server.ServerMethodThatReturnsTwoWayStream));
+
+        int s = await result.ReadAsync(buffer, 0, buffer.Length, this.TimeoutToken);
+        string returnedContent = Encoding.UTF8.GetString(buffer, 0, s);
+
+        Assert.Equal("Streamed bits!", returnedContent);
+    }
+
+    /// <summary>
+    /// Verify that an inner stream also gets it's own multiplexing channel.
+    /// </summary>
+    [Fact]
+    public async Task ServerMethodThatReturnsCustomTypeWithInnerStream()
+    {
+        byte[] buffer = new byte[32];
+        StreamContainingClass result = await this.clientRpc.InvokeAsync<StreamContainingClass>(nameof(Server.ServerMethodThatReturnsCustomTypeWithStream));
+
+        int s = await result.InnerStream.ReadAsync(buffer, 0, buffer.Length, this.TimeoutToken);
+        string returnedContent = Encoding.UTF8.GetString(buffer, 0, s);
+
+        Assert.Equal("More streamed bits!", returnedContent);
+    }
+
+    /// <summary>
+    /// Verify that when a user requests a stream and doesn't consume it, we dispose of it on the server.
+    /// </summary>
+    [Fact]
+    public async Task ClientCanRequestStreamAndDropIt()
+    {
+        byte[] buffer = new byte[32];
+        await this.clientRpc.InvokeAsync(nameof(Server.ServerMethodThatReturnsAndDisposesStream));
+
+        this.serverRpc.Dispose();
+
+        await WhenAllSucceedOrAnyFault(this.server.StreamDisposedTask!);
```

Comment:
I'm not sure what you're trying to show here. Of course this will complete because your server method scheduled its disposal before the client even had a chance to read from it.
I wonder if you were trying to test what I was talking about where closing the JsonRpc connection is supposed to dispose all active streams. I don't see a test here that covers that scenario. In that case you'd remove all explicit Stream disposals, but go ahead and dispose the client *or* server's `JsonRpc` object. Then verify that both ends of the `Stream` did in fact get disposed of.
The `JsonRpcClosesStreamAfterDisconnectedEvent` test looks like it may be helpful to at least look at here.